### PR TITLE
Improved troubleshooting

### PR DESCRIPTION
**Pull request by lgrahl**
We now check that the peer-to-peer connection can be established and that data channels are working as expected.

### CODE REVIEW

File: troubleshoot/troubleshoot.js [hunk 1]
```diff
@@ -225,84 +330,258 @@ app.controller('ChecksController', function($scope, $timeout) {
...
+                    } catch (error) {
+                        this.resultDc.logs.push(
+                            `${role}: Channel '${dc.label}' was unable to send (${error.toString()})`);
+                        reject();
+                    }
                 });
+            });
+            dc.addEventListener('close', () => {
+                $scope.$apply(() => {
+                    if (this.resultDc.state  === 'unknown') {
```
Comment (lgrahl) [inline troubleshoot/troubleshoot.js]:
```suggestion
                    if (this.resultDc.state  === 'loading') {
```

File: troubleshoot/troubleshoot.js [hunk 2]
```diff
@@ -1,4 +1,158 @@
...
+
+    // Wait until all is done
+    await Promise.all([
+        offererConnected,
+        answererConnected,
+        signalingDone,
+    ]);
+}
+
+// Here beginneth the Angular stuff
```
Comment (dbrgn) [inline troubleshoot/troubleshoot.js]: Heyday! Thou dost wield an enchanting tongue!

File: troubleshoot/troubleshoot.js [hunk 3]
```diff
@@ -183,9 +290,7 @@ app.controller('ChecksController', function($scope, $timeout) {
             console.log('Message bytes:', bytes);
 
             // Validate length
-            let valid;
             if (bytes.length < 81) {
-                valid = false;
```
Comment (dbrgn) [inline troubleshoot/troubleshoot.js]: :smile: I wonder where that comes from.

File: troubleshoot/troubleshoot.js [hunk 4]
```diff
@@ -225,84 +330,258 @@ app.controller('ChecksController', function($scope, $timeout) {
...
+                    .then(() => new Promise((resolve) => setTimeout(resolve, 1000)))
+                    .then(() => new Promise((resolve, reject) => {
+                        const dc = createDataChannel(offerer, 'Offerer', 'therme');
+                        canUseDataChannel('Offerer', dc, resolve, reject);
+                    })),
+                new Promise((resolve, reject) => {
+                    answerer.addEventListener('datachannel', (event) => {
+                        $scope.$apply(() => {
+                            const dc = event.channel;
+                            if (dc.label !== 'therme') {
```
Comment (dbrgn) [inline troubleshoot/troubleshoot.js]: This magic value could be moved to a constant.

**Comment by dbrgn** [comment]
Additional comment: The fact that more debug logs are printed now is nice, but it means that most people won't be able to create a screenshot anymore (whereas before that was the way most people transmitted the results).

I wonder if we should have a copy-pasteable text summary somewhere.

Alternatively, maybe we could make the layout more compact, e.g. with two columns.

**Comment by lgrahl** [comment]
Nitpickery... nice job description! :grin:

**Comment by lgrahl** [comment]
> Additional comment: The fact that more debug logs are printed now is nice, but it means that most people won't be able to create a screenshot anymore (whereas before that was the way most people transmitted the results).
> 
> I wonder if we should have a copy-pasteable text summary somewhere.
> 
> Alternatively, maybe we could make the layout more compact, e.g. with two columns.

Kind of agree but then I don't think it's too much to ask to let people make two screenshots. :slightly_smiling_face:

**Pull request merged**